### PR TITLE
Config provider

### PR DESCRIPTION
**Pull request by zombieJ**
Add a `ConfigProvider` for user to customize some global setting.

Currently let's add `getPopupContainer` first. `prefixCls` will be in another PR after this merged.

ref: #12985 #10855

### CODE REVIEW

File: components/layout/demo/side.md [hunk 1]
```diff
@@ -26,6 +26,8 @@ import { Layout, Menu, Breadcrumb, Icon, ConfigProvider } from 'antd';
 const { Header, Content, Footer, Sider } = Layout;
 const SubMenu = Menu.SubMenu;
 
+// 代码中 ConfigProvider 仅用于官网示例修正使用，你的代码并不需要用到 setContainerRef, getPopupContainer 和 ConfigProvider。删除即可。
+// `ConfigProvider` in sample is only for site demo fix usag, your real code needn't use setContainerRef, getPopupContainer & ConfigProvider. Just remove them.
```
Comment (ztplz) [inline components/layout/demo/side.md]: `usag` => `usage`

File: components/tooltip/index.tsx [hunk 2]
```diff
@@ -3,6 +3,7 @@ import { cloneElement } from 'react';
 import { polyfill } from 'react-lifecycles-compat';
 import RcTooltip from 'rc-tooltip';
 import classNames from 'classnames';
+import { ConfigConsumer, ConfigProviderProps } from '../config-provider';
 import getPlacements, { AdjustOverflow, PlacementsConfig } from './placements';
 import Button from '../button/index';
```
Comment (ilanus) [inline components/tooltip/index.tsx]: > ```diff
> - import { ConfigConsumer, ConfigProviderProps } from '../config-provider';
> - import getPlacements, { AdjustOverflow, PlacementsConfig } from './placements';
> - import Button from '../button/index';
> + import Button from '../button/index';
> + import { ConfigConsumer, ConfigProviderProps } from '../config-provider';
> + import getPlacements, { AdjustOverflow, PlacementsConfig } from './placements';
> ```

File: components/layout/__tests__/__snapshots__/demo.test.js.snap [hunk 3]
```diff
@@ -957,6 +957,7 @@ exports[`renders ./components/layout/demo/side.md correctly 1`] = `
   class="ant-layout"
   style="min-height:100vh"
 >
+  <div />
```
Comment (afc163) [inline components/layout/__tests__/__snapshots__/demo.test.js.snap]: 这是啥

**Comment by afc163** [comment]
补个用例？

**Comment by afc163** [comment]
是不是可以加入 message.config 和 notification.config 的功能。

File: components/cascader/index.tsx [hunk 4]
```diff
@@ -496,4 +499,12 @@ export default class Cascader extends React.Component<CascaderProps, CascaderSta
       </RcCascader>
     );
   }
+
+  render() {
+    return (
+      <ConfigConsumer>
+        {this.renderCascader}
```
Comment (ilanus) [inline components/cascader/index.tsx]: `this.renderCascader()`

File: components/config-provider/index.en-US.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+type: Other
+cols: 1
+title: ConfigProvider
+---
+
+`ConfigProvider` provides a uniform configuration support for components.
+
+## Usage
+
+`ConfigProvider` takes use of [context](https://facebook.github.io/react/docs/context.html), a feature of React, to accomplish global effectiveness by wrapping the app only once.
```
Comment (ilanus) [inline components/config-provider/index.en-US.md]: This component provides a configuration to all React components underneath itself via the [context API](https://facebook.github.io/react/docs/context.html), In the render tree all components will have access to the provided config.

File: components/cascader/index.tsx (hunk 4 above)
Comment (ilanus) [inline components/cascader/index.tsx]: 在哪里渲染 `()` ？

Comment (zombieJ) [reply components/cascader/index.tsx]: 这里：https://github.com/ant-design/ant-design/pull/12991/files/0f554a8e4e47ba084ef59e844ab0446ee585d0b3#diff-3141ae7cf791cd0b5edcad2c51dd6dddR352

**Comment by zombieJ** [comment]
merged feature branch. Pls check.

**Comment by ycjcl868** [comment]
test cases need to add.

**Pull request merged**